From c459dea18a6e57a2625f8f0ec1f7a0c259ab1d95 Mon Sep 17 00:00:00 2001
From: Fred Fan <r01011@freescale.com>
Date: Tue, 5 Aug 2008 16:29:40 +0800
Subject: [PATCH] ENGR00086569 The flexcan can not work as module on imx35_3stack

The root cause is flexcan does not declare as GPL license.
And sync the bitrate when the factors are changed.

Signed-off-by: Fred Fan <r01011@freescale.com>
---
 drivers/net/can/flexcan/dev.c |   17 ++++++++++++-----
 drivers/net/can/flexcan/drv.c |    4 ++--
 drivers/net/can/flexcan/mbm.c |    2 --
 3 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/net/can/flexcan/dev.c b/drivers/net/can/flexcan/dev.c
index 4cf2a21..79be30f 100644
--- a/drivers/net/can/flexcan/dev.c
+++ b/drivers/net/can/flexcan/dev.c
@@ -29,7 +29,6 @@
 #include <linux/device.h>
 
 #include <linux/module.h>
-#include <linux/version.h>
 #include "flexcan.h"
 
 enum {
@@ -351,24 +350,32 @@ static ssize_t flexcan_set_attr(struct device *dev,
 		flexcan_set_bitrate(flexcan, tmp);
 		break;
 	case FLEXCAN_ATTR_BR_PRESDIV:
-		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PRESDIV))
+		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PRESDIV)) {
 			flexcan->br_presdiv = tmp - 1;
+			flexcan_update_bitrate(flexcan);
+		}
 		break;
 	case FLEXCAN_ATTR_BR_RJW:
 		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_RJW))
 			flexcan->br_rjw = tmp - 1;
 		break;
 	case FLEXCAN_ATTR_BR_PROPSEG:
-		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PROPSEG))
+		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PROPSEG)) {
 			flexcan->br_propseg = tmp - 1;
+			flexcan_update_bitrate(flexcan);
+		}
 		break;
 	case FLEXCAN_ATTR_BR_PSEG1:
-		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PSEG1))
+		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PSEG1)) {
 			flexcan->br_pseg1 = tmp - 1;
+			flexcan_update_bitrate(flexcan);
+		}
 		break;
 	case FLEXCAN_ATTR_BR_PSEG2:
-		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PSEG2))
+		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_PSEG2)) {
 			flexcan->br_pseg2 = tmp - 1;
+			flexcan_update_bitrate(flexcan);
+		}
 		break;
 	case FLEXCAN_ATTR_MAXMB:
 		if ((tmp > 0) && (tmp <= FLEXCAN_MAX_MB)) {
diff --git a/drivers/net/can/flexcan/drv.c b/drivers/net/can/flexcan/drv.c
index 87f98d2..a65b626 100644
--- a/drivers/net/can/flexcan/drv.c
+++ b/drivers/net/can/flexcan/drv.c
@@ -22,8 +22,6 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/slab.h>
-#include <linux/module.h>
-#include <linux/init.h>
 #include <linux/netdevice.h>
 #include <linux/if_arp.h>
 #include <linux/if_ether.h>
@@ -620,3 +618,5 @@ static __exit void flexcan_exit(void)
 
 module_init(flexcan_init);
 module_exit(flexcan_exit);
+
+MODULE_LICENSE("GPL");
diff --git a/drivers/net/can/flexcan/mbm.c b/drivers/net/can/flexcan/mbm.c
index 7090721..1a3eeb3 100644
--- a/drivers/net/can/flexcan/mbm.c
+++ b/drivers/net/can/flexcan/mbm.c
@@ -21,8 +21,6 @@
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/slab.h>
-#include <linux/module.h>
-#include <linux/init.h>
 #include <linux/netdevice.h>
 #include <linux/if_arp.h>
 #include <linux/if_ether.h>
-- 
1.5.4.4

